### PR TITLE
Lazy loading

### PR DESCRIPTION
**Pull request by TheSlimvReal**
see issue: #1600 

### Visible/Frontend Changes
--

### Architectural/Backend Changes
- [x] All dynamic components are lazy loaded
- [x] `DemoDataModule` is lazy loaded
- [x] Most services are `providedIn: "root"` in order to support lazy loading
- [x] All components are standalone in order to support lazy loading
- [ ] Fix storybook stories

### CODE REVIEW

File: src/app/core/coming-soon/coming-soon/coming-soon.component.ts [hunk 1]
```diff
@@ -9,17 +9,31 @@ import {
...
+import { FontAwesomeModule } from "@fortawesome/angular-fontawesome";
+import { NgIf } from "@angular/common";
+import { RouteTarget } from "../../../app.routing";
 
 /**
  * Placeholder page to announce that a feature is not available yet.
  *
  * This integrates with analytics and allows user to explicitly request the feature.
  */
+@RouteTarget("ComingSoon")
```
Comment (TheSlimvReal) [inline src/app/core/coming-soon/coming-soon/coming-soon.component.ts:24]: Should we replace the `RouteTarget` decorator with the `DynamicComponent` decorator as it is has no special meaning any more. In fact, both the registries are not really needed anymore as the lookup can happen through the global `dynamicComponents` map, which holds the lazy import functions. The `DynamicComponent` decorator has the benefit, that after the first loading, the component can be used without another lazy loading. What performance impact this has - I am not sure.

File: src/app/dynamic-components.ts [hunk 2]
```diff
@@ -0,0 +1,488 @@
+import { Type } from "@angular/core";
+
+// TODO make a build script that looks for annotations and creates this
+export const dynamicComponents = new Map<string, () => Promise<Type<any>>>([
+  [
+    "Dashboard",
+    () =>
+      import("./core/dashboard/dashboard/dashboard.component").then(
```
Comment (TheSlimvReal) [inline src/app/dynamic-components.ts:8]: I think these functions are quite bad for the code coverage as none of them are executed in the tests.

File: src/app/core/view/dynamic-components/dynamic-component.directive.ts [hunk 3]
```diff
@@ -22,16 +25,23 @@ export class DynamicComponentDirective implements OnChanges {
...
 
-    const component = this.registry.get(this.appDynamicComponent.component);
+    let component: ComponentType<OnInitDynamicComponent>;
+    if (this.registry.has(this.appDynamicComponent.component)) {
+      component = this.registry.get(this.appDynamicComponent.component);
+    } else {
+      component = await dynamicComponents.get(
+        this.appDynamicComponent.component
+      )();
+    }
```
Comment (sleidig) [inline src/app/core/view/dynamic-components/dynamic-component.directive.ts:44]: if (almost) everything is lazy-loaded, do we stil have use for the registry at all?

**Review by sleidig — APPROVED** [review]
Tons of code but actually pretty clear and good :slightly_smiling_face:

File: src/app/child-dev-project/attendance/attendance.module.ts [hunk 4]
```diff
@@ -83,87 +33,11 @@ export class HorizontalHammerConfig extends HammerGestureConfig {
...
-    AddDayAttendanceComponent,
-    AttendanceManagerComponent,
-    ActivityAttendanceSectionComponent,
-    AttendanceWeekDashboardComponent,
-    GroupedChildAttendanceComponent,
-    AttendanceBlockComponent,
-    RecentAttendanceBlocksComponent,
-  ];
-}
+export class AttendanceModule {}
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/child-dev-project/attendance/attendance.module.ts:43]: same for entityRegistry, ChildrenModule, PwaInstallModule - if they only provide one or two services, wouldn't it be better to do that in a different way than through an empty module?

**Comment by sleidig** [comment]
We should mark the final commit as QualitätsMENTOR also.

**Pull request merged**